common/quotas/ownership_scaled_quota.go

```diff
@@ -0,0 +1,161 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+)
+
+type (
+	// OwnershipScaledQuota is a RateBurst implementation that scales the RPS and burst quotas linearly with the
+	// fraction of the total shards in the cluster owned by this host. The effective Rate and Burst are both multiplied
+	// by:
+	//
+	//	shardCount / totalShards * constantScaleFactor
+	//
+	// Note that there is no scaling, not even by the constantScaleFactor, until the first shard count update is
+	// received.
+	OwnershipScaledQuota struct {
+		// rb is the base quota that we will scale.
+		rb RateBurst
+		// shardCount is the number of shards owned by this host.
+		shardCount atomic.Int64
+		// totalShards is the total number of shards in the cluster.
+		totalShards int
+		// constantScaleFactor is a constant scale factor that is multiplied by the fraction of the total shards in the
+		// cluster owned by this host to get the final scale factor.
+		constantScaleFactor float64
```

Comment:
what's the purpose of this factor? are we going to set it to 1 for our use case?

Reply:
I went ahead and removed this though because it's not necessary